diff --git a/ts2panda/ts2abc/ts2abc.cpp b/ts2panda/ts2abc/ts2abc.cpp
index 1f3e8c12..d25e479f 100755
--- a/ts2panda/ts2abc/ts2abc.cpp
+++ b/ts2panda/ts2abc/ts2abc.cpp
@@ -402,6 +402,10 @@ static void ParseInstructionDebugInfo(const Json::Value &ins, panda::pandasm::In
             if (debugPosInfo.isMember("wholeLine") && debugPosInfo["wholeLine"].isString()) {
                 insDebug.whole_line = debugPosInfo["wholeLine"].asString();
             }
+
+            if (debugPosInfo.isMember("columnNum") && debugPosInfo["columnNum"].isInt()) {
+                insDebug.column_number = debugPosInfo["columnNum"].asInt();
+            }
         }
 
         if (debugPosInfo.isMember("lineNum") && debugPosInfo["lineNum"].isInt()) {
